### PR TITLE
refactor: refactor ScaledJob to be used with scale_handler

### PR DESCRIPTION
**Pull request by yoongon**
_Provide a description of what has been changed_
Currently, ScaledJob is left at scalers_cache.go. Move this one into the scale_handler.go for consistent with ScaledObject.

FYI.
https://github.com/kedacore/keda/blob/bce33756073562a462251dba7f2638aa0b393ca6/pkg/scaling/cache/scalers_cache.go#L144 

### Checklist

- [x] I have verified that my change is according to the [deprecations & breaking changes policy](https://github.com/kedacore/governance/blob/main/DEPRECATIONS.md)
- [x] Tests have been added
- [x] Commits are signed with Developer Certificate of Origin (DCO - [learn more](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#developer-certificate-of-origin-signing-your-work))

Fixes https://github.com/kedacore/keda/issues/4781

### CODE REVIEW

File: pkg/scaling/scale_handler.go [hunk 1]
```diff
@@ -611,3 +607,171 @@ func (h *scaleHandler) getScaledObjectState(ctx context.Context, scaledObject *k
...
+		isActive = true
+	}
+
+	maxValue = min(float64(scaledJob.MaxReplicaCount()), maxValue)
+	logger.V(1).WithValues("ScaledJob", scaledJob.Name).Info("Checking if ScaleJob Scalers are active", "isActive", isActive, "maxValue", maxValue, "MultipleScalersCalculation", scaledJob.Spec.ScalingStrategy.MultipleScalersCalculation)
+
+	return isActive, ceilToInt64(queueLength), ceilToInt64(maxValue)
+}
+
+func getTargetAverageValue(metricSpecs []v2.MetricSpec) float64 {
```
Comment (zroubalik) [inline pkg/scaling/scale_handler.go:748]: We should move this functions to a separate file/package.

File: pkg/scaling/scale_handler.go [hunk 2]
```diff
@@ -611,3 +607,171 @@ func (h *scaleHandler) getScaledObjectState(ctx context.Context, scaledObject *k
...
+}
+
+// / --------------------------------------------------------------------------- ///
+// / ----------             ScaledJob related methods               --------- ///
+// / --------------------------------------------------------------------------- ///
+
+// getScaledJobMetrics returns metrics for specified metric name for a ScaledJob identified by its name and namespace.
+// It could either query the metric value directly from the scaler or from a cache, that's being stored for the scaler.
+func (h *scaleHandler) getScaledJobMetrics(ctx context.Context, scaledJob *kedav1alpha1.ScaledJob) []scalerMetrics {
+	// TODO this loop should be probably done similar way the ScaledObject loop is done
```
Comment (zroubalik) [inline pkg/scaling/scale_handler.go:624]: could you please address this TODO?

File: pkg/scaling/scale_handler.go [hunk 3]
```diff
@@ -611,3 +607,171 @@ func (h *scaleHandler) getScaledObjectState(ctx context.Context, scaledObject *k
...
+
+		metricSpecs := s.GetMetricSpecForScaling(ctx)
+
+		// skip scaler that doesn't return any metric specs (usually External scaler with incorrect metadata)
+		// or skip cpu/memory resource scaler
+		if len(metricSpecs) < 1 || metricSpecs[0].External == nil {
+			continue
+		}
+
+		// TODO here we should probably loop through all metrics in a Scaler
```
Comment (zroubalik) [inline pkg/scaling/scale_handler.go:649]: could you please address this TODO?

**Review by zroubalik — COMMENTED** [review]
Thanks for tackling this, my ambition behind this is not just move these functions but also revise & clenup the code. Also some additional tests would be nice.

File: pkg/scaling/scale_handler.go (hunk 2 above)
Comment (yoongon) [reply pkg/scaling/scale_handler.go:624]: Hi @zroubalik 
Cause I don't know the context very well, could you elaborate what todo means? BTW, I've changed it like below. I guess it's not similar any more.
I changed it like the below one, but I haven't deleted the line.

before
https://github.com/kedacore/keda/pull/4707/files#diff-e15b5649231cd2fc16ca66df8b66cc1c705c0daf7007b23a9a017bf0948409c4L239

now
https://github.com/kedacore/keda/pull/4707/files#diff-7e9a23e26ddcd00c13026a9c8f9bedcb7414c4cf1fe36a3a8ff1f29258fec055R632

File: pkg/scaling/scale_handler.go (hunk 3 above)
Comment (yoongon) [reply pkg/scaling/scale_handler.go:649]: ditto.

I think it's looping through all metrics now.
Could you share more detail?

File: pkg/scaling/scale_handler.go (hunk 1 above)
Comment (yoongon) [reply pkg/scaling/scale_handler.go]: Thanks. I updated

**Comment by yoongon** [comment]
@zroubalik 
Could you check my comment again? Thanks.

File: pkg/scaling/scale_handler.go [hunk 4]
```diff
@@ -611,3 +607,152 @@ func (h *scaleHandler) getScaledObjectState(ctx context.Context, scaledObject *k
...
+				isActive = metrics.isActive
+			}
+		}
+	}
+
+	if scaledJob.MinReplicaCount() > 0 {
+		isActive = true
+	}
+
+	maxValue = min(float64(scaledJob.MaxReplicaCount()), maxValue)
```
Comment (zroubalik) [inline pkg/scaling/scale_handler.go:742]: could you please move this to a separate function in a package `pkg/scaling/scaledjob`?

File: pkg/scaling/scale_handler.go [hunk 5]
```diff
@@ -611,3 +607,152 @@ func (h *scaleHandler) getScaledObjectState(ctx context.Context, scaledObject *k
...
+
+	return isActive, ceilToInt64(queueLength), ceilToInt64(maxValue)
+}
+
+func ceilToInt64(x float64) int64 {
+	return int64(math.Ceil(x))
+}
+
+// Min function for float64
+func min(x, y float64) float64 {
```
Comment (zroubalik) [inline pkg/scaling/scale_handler.go:753]: Could you please move this to a separate function in a package pkg/scaling/scaledjob? And refactor the code a little bit so these don't have to be a public functions if possible.

File: pkg/scaling/scale_handler.go (hunk 3 above)
Comment (zroubalik) [reply pkg/scaling/scale_handler.go:649]: Ah, I can see it is fixed, then please remove the TODO

File: pkg/scaling/scale_handler.go (hunk 2 above)
Comment (zroubalik) [reply pkg/scaling/scale_handler.go:624]: Yeah, then please remove the TODO

File: pkg/scaling/utils/metrics.go [hunk 6]
```diff
@@ -0,0 +1,39 @@
+package utils
```
Comment (zroubalik) [inline pkg/scaling/utils/metrics.go:1]: Could you please rename this package to `scaledjob`

File: pkg/scaling/utils/metrics.go [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+
+import (
+	v2 "k8s.io/api/autoscaling/v2"
+	"k8s.io/apimachinery/pkg/api/resource"
+)
+
+func GetTargetAverageValue(metricSpecs []v2.MetricSpec) float64 {
```
Comment (zroubalik) [inline pkg/scaling/utils/metrics.go:8]: Pleaes add a comment with explanation (doc), thanks!

File: pkg/scaling/utils/metrics.go [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+		totalAverageValue += metricValue
+	}
+	count := float64(len(metricSpecs))
+	if count != 0 {
+		return totalAverageValue / count
+	}
+	return 0
+}
+
+func CreateMetricSpec(averageValue int64, metricName string) v2.MetricSpec {
```
Comment (zroubalik) [inline pkg/scaling/utils/metrics.go:27]: Pleaes add a comment with explanation (doc), thanks!

**Review by zroubalik — COMMENTED** [review]
@yoongon first of all, sorry for the delay, I have been busy with other stuff. Great job, could you please address the rest of the minor concerns? Thanks!

**Review by zroubalik — COMMENTED** [review]
@yoongon could you please also create an issue for this effort and link it in the Changelog? (Other section)

File: pkg/scaling/scale_handler.go (hunk 2 above)
Comment (yoongon) [reply pkg/scaling/scale_handler.go]: Done.

File: pkg/scaling/scale_handler.go (hunk 3 above)
Comment (yoongon) [reply pkg/scaling/scale_handler.go]: Done.

File: pkg/scaling/scale_handler.go (hunk 4 above)
Comment (yoongon) [reply pkg/scaling/scale_handler.go]: Done.

File: pkg/scaling/utils/metrics.go (hunk 6 above)
Comment (yoongon) [reply pkg/scaling/utils/metrics.go]: Done.

File: pkg/scaling/utils/metrics.go (hunk 7 above)
Comment (yoongon) [reply pkg/scaling/utils/metrics.go]: Done.

File: pkg/scaling/utils/metrics.go (hunk 8 above)
Comment (yoongon) [reply pkg/scaling/utils/metrics.go]: Done.

File: pkg/scaling/scale_handler.go (hunk 5 above)
Comment (yoongon) [reply pkg/scaling/scale_handler.go]: Done.

**Comment by yoongon** [comment]
@zroubalik
Thanks for the detail check. I've updated code based on your review and update changelog as well.
Please check and let me know if there is anything I can contribute more :)

File: pkg/scaling/scaledjob/metrics.go [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+	}
+	count := float64(len(metricSpecs))
+	if count != 0 {
+		return totalAverageValue / count
+	}
+	return 0
+}
+
+// CreateMetricSpec creates MetricSpec for given metric name and target value.
+func CreateMetricSpec(averageValue int64, metricName string) v2.MetricSpec {
```
Comment (zroubalik) [inline pkg/scaling/scaledjob/metrics.go:31]: Isn't this used only in the test? Then it doesn't make sense to define here imho.

File: pkg/scaling/scaledjob/metrics.go [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+
+	maxValue = GetMaxValue(maxValue, maxReplicaCount)
+	return isActive, ceilToInt64(queueLength), ceilToInt64(maxValue), maxValue
+}
+
+func ceilToInt64(x float64) int64 {
+	return int64(math.Ceil(x))
+}
+
+// Min function for float64
```
Comment (zroubalik) [inline pkg/scaling/scaledjob/metrics.go:111]: please follow Go convention:
```
// min returns the minimum for input float64 values
```

File: pkg/scaling/scaledjob/metrics.go [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+
+	if minReplicaCount > 0 {
+		isActive = true
+	}
+
+	maxValue = GetMaxValue(maxValue, maxReplicaCount)
+	return isActive, ceilToInt64(queueLength), ceilToInt64(maxValue), maxValue
+}
+
+func ceilToInt64(x float64) int64 {
```
Comment (zroubalik) [inline pkg/scaling/scaledjob/metrics.go:107]: add comment here as well

File: pkg/scaling/scaledjob/metrics.go [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+
+// Min function for float64
+func min(x, y float64) float64 {
+	if x > y {
+		return y
+	}
+	return x
+}
+
+func GetMaxValue(maxValue float64, maxReplicaCount int64) float64 {
```
Comment (zroubalik) [inline pkg/scaling/scaledjob/metrics.go:119]: add comment pls

File: pkg/scaling/scaledjob/metrics.go [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+	}
+}
+
+type ScalerMetrics struct {
+	QueueLength float64
+	MaxValue    float64
+	IsActive    bool
+}
+
+func IsScaledJobActive(scalersMetrics []ScalerMetrics, multipleScalersCalculation string, minReplicaCount, maxReplicaCount int64) (bool, int64, int64, float64) {
```
Comment (zroubalik) [inline pkg/scaling/scaledjob/metrics.go:51]: add comment

File: pkg/scaling/scale_handler.go [hunk 14]
```diff
@@ -620,3 +615,83 @@ func (h *scaleHandler) getScaledObjectState(ctx context.Context, scaledObject *k
...
+		scalerLogger.V(1).Info("Scaler Metric value", "isTriggerActive", isTriggerActive, metricSpecs[0].External.Metric.Name, queueLength, "targetAverageValue", targetAverageValue)
+
+		if isTriggerActive {
+			isActive = true
+		}
+
+		if targetAverageValue != 0 {
+			averageLength := queueLength / targetAverageValue
+			maxValue = scaledjob.GetMaxValue(averageLength, scaledJob.MaxReplicaCount())
+		}
```
Comment (zroubalik) [inline pkg/scaling/scale_handler.go:675]: Sorry to be nit picky, but could you please also move this to the new package? To some a function that returns `queueLength` and `maxValue`? Something like `func calculateQueueLengthAndMaxValue(...) (queueLenght, maxValue float64)` ?

Comment (zroubalik) [reply pkg/scaling/scale_handler.go:675]: The goal is to reduce the number of public functions within `scaledjob` package

**Review by zroubalik — COMMENTED** [review]
Thanks, there are some leftovers, sorry for being so pedantic about this 😄

File: CHANGELOG.md [hunk 15]
```diff
@@ -172,6 +172,7 @@ New deprecation(s):
 - **General**: Bump `kubernetes-sigs/controller-runtime` to v0.15.0 and code alignment ([#4582](https://github.com/kedacore/keda/pull/4582))
 - **General**: Drop a transitive dependency on bou.ke/monkey ([#4364](https://github.com/kedacore/keda/issues/4364))
 - **General**: Fix odd number of arguments passed as key-value pairs for logging ([#4368](https://github.com/kedacore/keda/issues/4368))
+- **General**: Refactor ScaledJob related methods to be located at scale_handler  ([#4781](https://github.com/kedacore/keda/issues/4781))
```
Comment (zroubalik) [inline CHANGELOG.md:175]: This should be in the Unreleased section, not in 2.11

Comment (yoongon) [reply CHANGELOG.md]: Nice catch. I made a mistake.
Done.

File: pkg/scaling/scale_handler.go (hunk 14 above)
Comment (yoongon) [reply pkg/scaling/scale_handler.go]: For more detail, I made **GetTargetAverageValue** a private. Please check.
Done.

File: pkg/scaling/scaledjob/metrics.go (hunk 13 above)
Comment (yoongon) [reply pkg/scaling/scaledjob/metrics.go:70]: Done.

File: pkg/scaling/scaledjob/metrics.go (hunk 12 above)
Comment (yoongon) [reply pkg/scaling/scaledjob/metrics.go]: Done.

File: pkg/scaling/scaledjob/metrics.go (hunk 10 above)
Comment (yoongon) [reply pkg/scaling/scaledjob/metrics.go]: I updated comment. Thanks.
Done.

File: pkg/scaling/scaledjob/metrics.go (hunk 9 above)
Comment (yoongon) [reply pkg/scaling/scaledjob/metrics.go:49]: I really appreciate your comment.
BTW, this function is used at scaler_handler_test.go as well., so I left here.
Could you give any suggestion?

Comment (zroubalik) [reply pkg/scaling/scaledjob/metrics.go:49]: So would be better to move to the test file and make it private?

Comment (yoongon) [reply pkg/scaling/scaledjob/metrics.go:49]: Hi, what I mean is,
it's been used two different package(scaledjob/metric_test.go, scaling/scaler_handler_test.go). If i make it private, I need to write duplicated code at both. 
Any good convention for this case?

Comment (zroubalik) [reply pkg/scaling/scaledjob/metrics.go:49]: That's right. I think that we can duplicate the function in this case. Moving it to some another package would bring to much complexity.

Comment (yoongon) [reply pkg/scaling/scaledjob/metrics.go]: That sounds much reasonable.
I updated it.
Thanks for the detail.

**Review by zroubalik — APPROVED** [review]
LGTM

Thanks for the effort! Really appreciate this :)

**Pull request merged**